keras_cv/training/simclr/simclr_trainer.py

```diff
@@ -0,0 +1,189 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+from keras_cv.layers import preprocessing
+
+
+class SimCLRTrainer(keras.Model):
+    """Creates a self-supervised SimCLR trainer for a model.
+
+    References:
+        - [SimCLR paper](https://arxiv.org/pdf/2002.05709)
+
+    Args:
+        encoder: a `keras.Model` to be pre-trained. In most cases, this encoder
+            should not include a top dense layer.
+        include_probe: Whether to include a single fully-connected layer during
+            training for probing classification accuracy using the learned encoding.
+            Note that this should be specified iff training with labeled images.
+            If provided, `classes` must be provided.
+        classes: optional number of classes to classify images into, only to be
+            specified if `include_top` is True.
+        projection_width: an optional width for the projection MLP in the SimCLR architecture
+        augmenter: a preprocessing layer to randomly augment input images for contrastive learning.
+            If no custom augmenter is provided, a default augmenter will be used,
+            and `value_range` must be provided.
+        value_range: the range of values the incoming images will have.
+            Required iff no `augmenter` is provided.
+            Represented as a two number tuple written [low, high].
+            This is typically either `[0, 1]` or `[0, 255]` depending
+            on how your preprocessing pipeline is setup.
+
+    Returns:
+      A `keras.Model` instance.
+
+    """
+
+    def __init__(
+        self,
+        encoder,
+        include_probe,
+        classes=None,
+        projection_width=128,
```

Comment:
Instead of taking `projection_width` as a parameter I'd rather take `projection_head` as a parameter to support all use cases